### PR TITLE
Adds Auto ItemsPerPage (TABLE/PAGINATOR) GH-145

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @jbrundage @anmoljagetia Please Review

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

File: src/other/Table.js [hunk 1]
```diff
@@ -96,22 +96,59 @@
             .remove()
         ;
 
+        var theadHeight = parseInt(th.style('height'));
+        var theadMarginPadding = parseInt(th.style('padding-top')) + parseInt(th.style('padding-bottom')) + parseInt(th.style('margin-top')) + parseInt(th.style('margin-bottom'));
```
Comment (GordonSmith) [inline src/other/Table.js]: If you look in HTMLWidget there are some calcHeight functions which do this

**Comment by GordonSmith** [comment]
I would suggest moving the entire "calcHeight" piece into its own method (like we do in other cases).

**Comment by GordonSmith** [comment]
The other thought I had was within the calcHeight function, you could simply create a new instance of the TableWidget - render it with no rows, measure height, render with one row, measure height, destroy widget.

**Comment by mlzummo** [comment]
can you show an example of where we move the calcHeight blocks of code into there own functions... i think using this.calcHeight did the trick enough?

**Comment by GordonSmith** [comment]
If you add a new method called "calcRowsPerPage" or such like and them move all you calculation code into that and just call it once, it would be cleaner - XYAxis does something similar:  https://github.com/hpcc-systems/Visualization/blob/master/src/chart/XYAxis.js#L58-L80

**Comment by mlzummo** [comment]
@GordonSmith made suggested changes ... Please Review

**Comment by GordonSmith** [comment]
If you rollup the two commits I can comment on the individual lines.

There is one console.log that can be removed.

I suspect that if and empty data set is passed in your tmpCell will be displayed?  I think you should "remove" it once you have measured its height?

Also I would have expected all the measurement code to go into the other a diff method to keep the main update call clean.

**Comment by mlzummo** [comment]
@GordonSmith ill take some suggestions around how to make the code leaner/cleaner, this is the structure i came up with but if there is a better approach can you provide some pseudo code around a better method of keeping the update call clean?

**Comment by GordonSmith** [comment]
What I was trying to get at would have been (in the update method):  var maxNumRows = this.calcMaxRows(x, y, z, ...);

**Comment by mlzummo** [comment]
@GordonSmith Please Review ... made suggested changes

**Comment by mlzummo** [comment]
it should auto remove the tcell as well even though i don't explicitly call .remove() on it due to later updates?

**Comment by GordonSmith** [comment]
Looks good.

**Pull request merged**